### DISCUSSION TITLE
Getting WARNING logs related to kafka sink

### DISCUSSION THREAD

**Discussion by ShahroZafar** [vectordotdev/vector#20510] [Q&A] [answered]
I am getting these logs for kafka sink which doesn't make sense. Does anybody know about them? 
I am on vector version 0.38.0
```
2024-05-17T06:16:51.858517Z  WARN librdkafka: librdkafka: CONFWARN [thrd:app]: Configuration property queue.buffering.max.ms is a producer property and will be ignored by this consumer instance
2024-05-17T06:16:51.858546Z  WARN librdkafka: librdkafka: CONFWARN [thrd:app]: Configuration property batch.size is a producer property and will be ignored by this consumer instance
2024-05-17T06:16:51.858550Z  WARN librdkafka: librdkafka: CONFWARN [thrd:app]: Configuration property request.required.acks is a producer property and will be ignored by this consumer instance
2024-05-17T06:16:51.858552Z  WARN librdkafka: librdkafka: CONFWARN [thrd:app]: Configuration property message.timeout.ms is a producer property and will be ignored by this consumer instance
```

**Comment by jszwedko** [comment]
Can you share your configuration?

**Reply by ShahroZafar** [reply]
Configurations are as follows
```
sources:
    kubernetes_logs:
      type: kubernetes_logs
      glob_minimum_cooldown_ms: 2000
      ingestion_timestamp_field: "ingest_timestamp"
      include_paths_glob_patterns:
        - /var/log/pods/istio-ingress_*/**
      namespace_annotation_fields: 
        namespace_labels: ""
      node_annotation_fields:
        node_labels: ""
      pod_annotation_fields:
        container_id: ""
        container_image_id: ""
    internal_metrics:
      type: internal_metrics
  sinks:
    prom_exporter:
      type: prometheus_exporter
      flush_period_secs: 30
      inputs: 
        - internal_metrics
      address: 0.0.0.0:9090
      buffer:
        type: disk
        when_full: block
        max_size: 4000000000
    kafka:
      type: kafka
      inputs:
        - kubernetes_logs
      bootstrap_servers: brokers:9092
      topic: vector-ingress-logs
      encoding:
        codec: json
      compression: zstd
      healthcheck:
        enabled: true
      batch:
        timeout_secs: 0.2
      librdkafka_options:
        client.id: "vector"
        request.required.acks: "1"
        message.timeout.ms: "0"
        linger.ms: "100"
        batch.size: "1000000"
      message_timeout_ms: 0
      buffer:
        type: disk
        when_full: block
```

**Reply by jszwedko** [reply]
Thanks! that is weird. The sink _is_ a producer instance and so I wouldn't expect those warnings. I ran this locally and saw the same.

**Reply by ShahroZafar** [reply]
I wonder if these warning logs are just to the extent of logs or `librdkafka_options` doesn't take effect at all

**Reply by jszwedko** [reply]
Oh I see what's happening here. The "healthcheck" is run by a Kafka consumer instance, but the rdkafka options are also applied ot it (in addition to the producer instance for sending messages). That is definitely confusing. I think ideally the logs would make that clearer.

**Reply by ShahroZafar** [reply]
So if I understand correctly, 2 instances are created consumer and producers?

consumer -> for healthchecks, librdkafka_options ignored
producer -> pushing logs to kafka and librdkafka_options are indeed applied?

**Reply by jszwedko** [reply, marked as answer]
Well, for healthchecks, only options that only apply to consumers are ignored. There are some options that can apply to both consumers and producers that will still apply to healthchecks. For the producer, you are correct.

**Comment by belltoy** [comment]
I have the same question about this warning.

After trace back the commit history, I wonder why the original commit for healthcheck of Kafka sink use a _consumer_ instead of a lower level client back in 2019 (b8de832b904b647c25b0f880171fc9fa49f5f7d3). I guess the reason is the high level producer (Future and Threaded) in `rdkafka` doesn't provide fetch_metadata API.

Another question is why the `KafkaRole` introduced in the commit 69c398a. Was it to distinguish the Consumer usage in the `healthcheck` function? If it was, I think the code of healthcheck in the Kafka sink can be simplified.

**Reply by jszwedko** [reply]
> After trace back the commit history, I wonder why the original commit for healthcheck of Kafka sink use a _consumer_ instead of a lower level client back in 2019 ([b8de832](https://github.com/vectordotdev/vector/commit/b8de832b904b647c25b0f880171fc9fa49f5f7d3)). I guess the reason is the high level producer (Future and Threaded) in `rdkafka` doesn't provide fetch_metadata API.

It would seem sensible to me to just fetch the metadata in the healthcheck to verify connectivity. It is actually available via https://docs.rs/rdkafka/latest/rdkafka/client/struct.Client.html#method.fetch_metadata. The client is available on `FutureProducer`: https://docs.rs/rdkafka/latest/rdkafka/producer/trait.Producer.html#tymethod.client. I'd support a PR for that.

> Another question is why the `KafkaRole` introduced in the commit [69c398a](https://github.com/vectordotdev/vector/commit/69c398a5b36067b4318ab552cf7363eb93bfe19f). Was it to distinguish the Consumer usage in the `healthcheck` function? If it was, I think the code of healthcheck in the Kafka sink can be simplified.

Agreed, I think it could be simplified if we avoid using a Kafka consumer in the sink.

Thanks for the thoughts @belltoy ! Let me know if you'd be interested in contributing these proposed changes.

**Reply by belltoy** [reply]
I'd like to commit a PR (#21129) for that.
